### ISSUE TITLE
How can I use training scripts of this repo to train the model on Google TPU?

### ISSUE THREAD

**Issue by AmitChaulwar** [UKPLab/sentence-transformers#672]
I want to train a QA model on MSMarco dataset from scratch using the script "train_biencoder.py". However, it requires huge time to train on GPU. I was wondering if there is quick example of how to train on TPU? Thanks.

**Comment by nreimers** [comment]
TPU training is not yet supported.

Note: You don't have to train the model until the end. Often you see after a fraction of the available data no further improvements on the development set. You can  then stop the training.

**Comment by AmitChaulwar** [comment]
Is there any plan for TPU support in near future?

**Comment by nreimers** [comment]
I am sadly afraid that I will not be able to add TPU support soon.

But a Pull Request adding TPU support would be highly appreciated.

**Comment by aditya-malte** [comment]
@nreimers With huggingface accelerate it should be pretty easy now as shown [here](https://huggingface.co/blog/accelerate-library) .
Let me know what you think, I’d be ready to work on it (adding support for TPU/multi-GPU)

**Comment by nreimers** [comment]
Hi @aditya-malte 
Yes, it is planned to integrate accelerate.

However, for TPU support, some more changes are needed: Currently the code uses dynamic padding, i.e. batches are padded to the length of the longest sentence in a batch. So every batch has a different length.

However, TPU require that you work with batches that all have the same length, i.e., the padding for TPU must be changed so that a fixed padding is used.

I think the changes for the CrossEncoder should be easy. For the BiEncoder (SentenceTransformer), it is a bit more complicated.
Several of the loss functions compute cossim across all embeddings in a batch. For a TPU, the individual embeddings must be computed on the individual cores of the TPU. Then, these must be moved back to a single core where the cossim computation can happen. 

I would be happy to have support here. If you like, you can maybe start with adding accelerate to the CrossEncoder training and create a pull request? After that, accelerate could be added to the BiEncoder (which is more complicated).

**Comment by aditya-malte** [comment]
Great! For cross-encoder it should be relatively easy then. By the way, does that mean that multi-GPU(and not TPU) training can be supported today itself for sentence transformers?

**Comment by nreimers** [comment]
So far multi GPU is also not supported. I tried the integration of DataParallel, but it was slower than training on a single GPU

**Comment by zbrnwpu** [comment]
Does the current model support multi-gpu training?@nreimers

**Comment by nreimers** [comment]
@zbrnwpu Currently not, will be part of the next release

**Comment by Desperado721** [comment]
Hi @nreimers, thank you for your suggestion about running model on TPU, however I changed the device to be xm.xla_device(), and put model on this device. I feel like the model is still trained on CPU. Could you help me with that? Thank you
